crates/symbolicator-service/src/services/sourcemap_lookup.rs

```diff
@@ -255,36 +285,218 @@ impl SourceMapLookup {
     ) -> Option<&CacheEntry<OwnedSourceMapCache>> {
         self.sourcemaps.get(abs_path)
     }
+
+    /// Fetches an arbitrary file using its `abs_path`,
+    /// or optionally its [`DebugId`] and [`SourceFileType`]
+    /// (because multiple files can share one [`DebugId`]).
+    pub async fn get_file(&mut self, key: FileKey) -> CacheEntry<CachedFile> {
+        // First, we do a trivial lookup in case we already have this file.
+        // The `abs_path` uniquely identifies a file in a JS stack trace, so we use that as the
+        // lookup key throughout.
+        if let Some(file) = self.files_by_key.get(&key) {
+            return file.clone();
+        }
+
+        // Otherwise, try looking it up in one of the artifact bundles that we already have.
+        if let Some(file) = self.try_get_file_from_bundles(&key) {
+            return self.files_by_key.entry(key).or_insert(Ok(file)).clone();
+        }
+
+        // Otherwise, we try to get the file from the already fetched existing individual files.
+
+        // TODO:
+        // look up every candidate `abs_path` in `self.artifacts`
+
+        // TODO:
+        // Otherwise, do a (cached) API request to Sentry and get the artifacts
+        //
+        //
+        // NOTE: We have `self.remote_artifacts` for the API requests already.
+
+        // TODO:
+        // Otherwise, fall back to scraping from the Web.
+
+        Err(CacheError::NotFound)
+    }
+
+    /// Gets the [`OwnedSourceMapCache`] for the file identified by its `abs_path`,
+    /// or optionally its [`DebugId`].
+    pub async fn get_sourcemapcache(&mut self, key: FileKey) -> CacheEntry<OwnedSourceMapCache> {
+        // First, check if we have already cached / created the `SourceMapCache`.
+        if let Some(smcache) = self.sourcemaps_by_key.get(&key) {
+            return smcache.clone();
+        }
+
+        // Fetch the minified file first
+        let minified_id = key
+            .debug_id
+            .map(|id| (id.0, SourceFileType::MinifiedSource));
+        let minified_key = FileKey {
+            abs_path: key.abs_path.clone(),
+            debug_id: minified_id,
+        };
+        let minified_source = self.get_file(minified_key).await?;
+
+        // Then fetch the corresponding sourcemap if we have a sourcemap reference
+        let sourcemap_id = key
+            .debug_id
+            .as_ref()
+            .map(|id| (id.0, SourceFileType::SourceMap));
+        let sourcemap_url = match minified_source.source_mapping_url.as_deref() {
+            Some(sourcemap_url) => {
+                // FIXME: if we have a data URL, we don’t need the base at all
```

Comment:
Hm, this is indeed a bit of a flaw in the API of `SourceMapUrl`. I can't think of a better way to do it, though.

Reply:
Maybe making `FileKey` into a proper enum covering the three cases that its doc comment mentions might be the way to go?

Reply:
Although, that would only cover the case that "we should have an `abs_path` in this case (minified source)".
The other thing is "we don’t even need the `abs_path` in this case (we have a data-url)" which it would not solve.

Reply:
Side note: I assume the reason this is written as a match and not`and_then` is so that you can early-return from the function with `?`. That is a super annoying wart in the language IMO.

Reply:
> Although, that would only cover the case that "we should have an `abs_path` in this case (minified source)". The other thing is "we don’t even need the `abs_path` in this case (we have a data-url)" which it would not solve.

I think this is acceptable. As you say, in the case of a minified file the `abs_path` is definitely available.